### PR TITLE
feat: add obs csi controller api

### PR DESCRIPTION
**Pull request by leleyi**
**What type of PR is this?**
 /kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/obs/controllerserver.go [hunk 1]
```diff
@@ -120,7 +142,35 @@ func (cs *controllerServer) ControllerUnpublishVolume(_ context.Context, _ *csi.
 
 func (cs *controllerServer) ListVolumes(_ context.Context, req *csi.ListVolumesRequest) (
 	*csi.ListVolumesResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+	log.Infof("ListVolumes: called with args %v", protosanitizer.StripSecrets(*req))
+
+	if req.MaxEntries < 0 {
+		return nil, status.Error(codes.InvalidArgument,
+			fmt.Sprintf("Validation failed, max entries request %v, must not be negative ", req.MaxEntries))
```
Comment (chengxiangdong) [inline pkg/obs/controllerserver.go:149]:
```suggestion
		return nil, status.Errorf(codes.InvalidArgument,
			"Validation failed, max entries request %v, must not be negative ", req.MaxEntries)
```

File: pkg/obs/controllerserver.go [hunk 2]
```diff
@@ -105,7 +106,28 @@ func (cs *controllerServer) DeleteVolume(_ context.Context, req *csi.DeleteVolum
...
-	return nil, status.Error(codes.Unimplemented, "")
+	log.Infof("ControllerGetVolume: called with args %v", protosanitizer.StripSecrets(*req))
+
+	volumeID := req.GetVolumeId()
+	if len(volumeID) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Validation failed, volume ID cannot be empty")
+	}
+
+	credentials := cs.Driver.cloud
+	bucket, err := services.GetParallelFSBucket(credentials, volumeID)
```
Comment (chengxiangdong) [inline pkg/obs/controllerserver.go:117]:
```suggestion
	bucket, err := services.GetParallelFSBucket(cs.Driver.cloud, volumeID)
```
`credentials` is only used once, no variable needs to be defined.

File: pkg/obs/controllerserver.go [hunk 3]
```diff
@@ -120,7 +142,35 @@ func (cs *controllerServer) ControllerUnpublishVolume(_ context.Context, _ *csi.
...
 	*csi.ListVolumesResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+	log.Infof("ListVolumes: called with args %v", protosanitizer.StripSecrets(*req))
+
+	if req.MaxEntries < 0 {
+		return nil, status.Error(codes.InvalidArgument,
+			fmt.Sprintf("Validation failed, max entries request %v, must not be negative ", req.MaxEntries))
+	}
+
+	volumes, err := services.ListBuckets(cs.Driver.cloud)
```
Comment (chengxiangdong) [inline pkg/obs/controllerserver.go:152]: Paginated queries should be used instead of querying all.

**Review by chengxiangdong — COMMENTED** [review]
`ListVolumes` need to be optimized，other parts are fine.

**Comment by chengxiangdong** [comment]
pls squash commits

File: pkg/obs/controllerserver.go [hunk 4]
```diff
@@ -120,7 +142,35 @@ func (cs *controllerServer) ControllerUnpublishVolume(_ context.Context, _ *csi.
...
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	entries := make([]*csi.ListVolumesResponse_Entry, 0, len(volumes))
+	for _, vol := range volumes {
```
Comment (deer-hang) [inline pkg/obs/controllerserver.go:158]: pls use MaxEntries and StartingToken processing the results

Comment (leleyi) [reply pkg/obs/controllerserver.go:161]: done

File: pkg/obs/controllerserver.go (hunk 3 above)
Comment (leleyi) [reply pkg/obs/controllerserver.go]: done

File: pkg/obs/controllerserver.go (hunk 2 above)
Comment (leleyi) [reply pkg/obs/controllerserver.go:117]: done

File: pkg/obs/controllerserver.go (hunk 1 above)
Comment (leleyi) [reply pkg/obs/controllerserver.go:149]: done

File: pkg/obs/services/buckets.go [hunk 5]
```diff
@@ -20,6 +20,7 @@ import (
 	"fmt"
 	"github.com/huaweicloud/huaweicloud-csi-driver/pkg/config"
 	"github.com/huaweicloud/huaweicloud-sdk-go-obs/obs"
+	"golang.org/x/sync/errgroup"
```
[the remark this answers is not in the document]
Comment (leleyi) [reply pkg/obs/services/buckets.go:23]: done

**Pull request merged**